### PR TITLE
#5544: Add output tensors parameter to  moreh_linear op

### PR DESCRIPTION
**Pull request by dongjin-na**
I added output tensor parameters to these OPs below refer to issue https://github.com/tenstorrent-metal/tt-metal/issues/5044 .

- moreh_linear
- moreh_linear_backward

Passed the relevant test below.

- tests/tt_eager/python_api_testing/unit_testing/misc/test_moreh_linear.py

### CODE REVIEW

**Review by eyonland — APPROVED** [review]
Consider using operation::ProgramWithOptionalOutputTensors as mentioned here:
https://tenstorrent-metal.github.io/tt-metal/latest/ttnn/ttnn/dependencies/tt_lib.html#new-device-operation-with-optional-output-tensors

**Comment by dongjin-na** [comment]
@mywoodstock , @TT-BrianLiu , @tt-aho
dear code owners, would you please review this PR? 
This PR is the same as https://github.com/tenstorrent/tt-metal/pull/7101

File: tests/tt_eager/python_api_testing/unit_testing/misc/test_moreh_linear.py [hunk 1]
```diff
@@ -156,10 +144,76 @@ def test_moreh_linear_backward(shapes, requires_grads, requires_bias_grad, devic
...
+        ([1, 1, 31, 31], [1, 1, 30, 31], [1, 1, 1, 1], [1, 1, 31, 30]),
+        ([2, 4, 4, 1024], [1, 1, 1536, 1024], [1, 1, 1, 1536], [2, 4, 4, 1536]),
+        ([1, 1, 32, 1023], [1, 1, 1536, 1023], [1, 1, 1, 1], [1, 1, 32, 1536]),
+    ),
+)
+@skip_for_wormhole_b0("disabled due to watcher error, see issue #5868")
+def test_moreh_linear_backward_enable_cache(shapes, device):
+    torch.manual_seed(3072)
+    requires_input_grad, requires_weight_grad, requires_bias_grad = (True, True, True)
+    device.enable_program_cache()
```
Comment (TT-BrianLiu) [inline tests/tt_eager/python_api_testing/unit_testing/misc/test_moreh_linear.py:213]: Use `use_program_cache` fixture?

File: tt_eager/tt_dnn/op_library/moreh_linear/moreh_linear_op.cpp [hunk 2]
```diff
@@ -23,43 +23,59 @@ inline bool is_shape_scalar(const Shape& bias) {
...
+    if (output.has_value()) {
+        const auto& output_tensor = output.value();
+        const auto& output_shape = output_tensor.get_legacy_shape().without_padding();
+        const auto& input_shape = input.get_legacy_shape().without_padding();
+        TT_ASSERT(
+            input_shape[0] == output_shape[0] &&
+            input_shape[1] == output_shape[1] &&
+            input_shape[2] == output_shape[2] &&
+            weight_shape[2] == output_shape[3],
+            "shape of output should be [input_shape[0], input_shape[1], input_shape[2], wieght_shape[2]]");
```
Comment (TT-BrianLiu) [inline tt_eager/tt_dnn/op_library/moreh_linear/moreh_linear_op.cpp:47]: spelling

**Review by TT-BrianLiu — APPROVED** [review]
I left some minor comments. Please address. Looks good to merge overall.

File: tests/tt_eager/python_api_testing/unit_testing/misc/test_moreh_linear.py (hunk 1 above)
Comment (dongjin-na) [reply tests/tt_eager/python_api_testing/unit_testing/misc/test_moreh_linear.py:213]: Okay. Applied it.

File: tt_eager/tt_dnn/op_library/moreh_linear/moreh_linear_op.cpp (hunk 2 above)
Comment (dongjin-na) [reply tt_eager/tt_dnn/op_library/moreh_linear/moreh_linear_op.cpp:47]: Thanks.

File: tt_eager/tt_dnn/op_library/moreh_linear/moreh_linear_op.cpp [hunk 3]
```diff
@@ -23,43 +23,59 @@ inline bool is_shape_scalar(const Shape& bias) {
...
-    std::optional<std::reference_wrapper<const Tensor>> bias,
+    const std::optional<const Tensor>& bias,
+    std::optional<Tensor> output,
     const MemoryConfig& output_mem_config) {
-    moreh_linear_validate(input, weight, bias);
-    Tensor mm_output = moreh_matmul(input, weight, std::nullopt, false, true, output_mem_config);
-    if (bias) {
-        const auto& bias_tensor = bias->get();
+    moreh_linear_validate(input, weight, bias, output);
+    output = moreh_matmul(input, weight, output, false, true, output_mem_config);
```
[the remark this answers is not in the document]
Comment (dongjin-na) [reply tt_eager/tt_dnn/op_library/moreh_linear/moreh_linear_op.cpp:58]: moreh_matmul returns a Tensor object, so there's no case that output is nullopt.
Thanks for the comment.

**Comment by dongjin-na** [comment]
all post-commit is passed. https://github.com/tenstorrent/tt-metal/actions/runs/8832482291

**Pull request merged**